### PR TITLE
feat: Rate-limit a warning about messages over a rate-limit

### PR DESCRIPTION
**Pull request by ukstv**
- Use PQueue for scheduling.
- Put a warning only if enough time has passed since the last warning.

### CODE REVIEW

**Review by stbrody — COMMENTED** [review]
A few small nitpicks on the tests to make sure they are clear.  Overall looks good though.  I'm a little saddened to be losing the ability to mock out time in another place and slow down our unit tests even further, but this is a problem in several places throughout our code at the moment and likely needs a more concerted effort to address than we can afford right now.

File: packages/core/src/dispatcher.ts [hunk 1]
```diff
@@ -170,7 +170,7 @@ export class Dispatcher {
    */
   async retrieveFromIPFS(cid: CID | string, path?: string): Promise<any> {
     try {
-      return this._getFromIpfs(cid, path)
+      return await this._getFromIpfs(cid, path)
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/dispatcher.ts:173]: Yes. No `await` means the try-catch block does nothing.

File: packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts [hunk 2]
```diff
@@ -32,146 +44,154 @@ describe('pubsub with queries rate limited', () => {
...
+      expect(
+        chunk.slice(1).every((timestamp) => Math.floor(timestamp.valueOf() - first) < ONE_SECOND)
+      ).toBeTruthy()
+    }
+  })
+
+  test('max number of queued queries', async () => {
+    const numMessages = MAX_QUEUED_QUERIES * 2
+    const original = vanillaPubsub.next.bind(vanillaPubsub)
+    vanillaPubsub.next = (message: QueryMessage) => {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts:121]: Ack. Remnant of copy-pasting other test.

File: packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts [hunk 3]
```diff
@@ -32,146 +44,154 @@ describe('pubsub with queries rate limited', () => {
...
-    mockNow.mockReturnValue(new Date(startTime.getTime() + 5000))
-    await allQueriesPublished
-    expect(ipfs.pubsub.publish).toBeCalledTimes(numMessages)
+    // Wait until the messages are submitted
+    // No errors expected
+    await Promise.all(messages.map((message) => whenSubscriptionDone(pubsub.next(message))))
 
-    mockNow.mockRestore()
+    eventsSubscription.unsubscribe()
+    expect(queueSizes.every((s) => s <= MAX_QUEUED_QUERIES)).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts:140]: Ack. With a comment.

File: packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts [hunk 4]
```diff
@@ -32,146 +44,154 @@ describe('pubsub with queries rate limited', () => {
...
+        QUERIES_PER_SECOND,
+        rateLimitWarningsIntervalMs
+      )
+
+      const warnMock = jest.fn(() => {
+        return new Subscription()
       })
+      diagnosticsLogger.warn = warnMock
+      // For clarity of the test, we disable processing here
+      pubsub.queue.pause()
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts:182]: If we don't, the queue processes the messages every second. It becomes harder to reason about timing.

Now with the queue paused, every message gets into the queue. It simulates the case of sending speed > rate limit.

File: packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts [hunk 5]
```diff
@@ -32,146 +44,154 @@ describe('pubsub with queries rate limited', () => {
...
+      // One call here
+      const subscriptions = messages.slice(0, 6).map((message) => pubsub.next(message))
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts]: Added the comment. Replaced to `QUERIES_PER_SECOND + 1`.

File: packages/core/src/pubsub/pubsub-ratelimit.ts [hunk 6]
```diff
@@ -21,46 +32,55 @@ export class PubsubRateLimit
...
   ) {
     super((subscriber) => {
       pubsub.subscribe(subscriber)
     })
 
-    this._queryQueue = new TaskQueue((err) => {
-      this.logger.err(`Error while publishing pubsub QUERY message: ${err}`)
+    // Limit number of executions by +intervalCap+ in +interval+ milliseconds.
+    // Here it is +queriesPerSecond+ per 1000ms = 1 second.
+    this.queue = new PQueue({ interval: 1000, intervalCap: queriesPerSecond })
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/pubsub/pubsub-ratelimit.ts:68]: It processes the tasks eagerly. Demonstrated here: https://github.com/ceramicnetwork/js-ceramic/pull/2424/files#diff-d8c717e006341f43804d1d58e6336aaf39a855efa28c653ec7b5177c88ab5890R144

File: packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts [hunk 7]
```diff
@@ -32,146 +44,154 @@ describe('pubsub with queries rate limited', () => {
...
-                }
-                if (i == numMessages) {
-                  resolveAllQueriesPublished()
-                }
-              })
-            })
-          }
-        )
-      })
+    vanillaPubsub.next = jest.fn(() => {
```
[the remark this answers is not in the document]
Comment (ukstv) [reply packages/core/src/pubsub/__tests__/pubsub-ratelimit.test.ts:86]: PQueue treats intervals optimistically. It does not assume that after the timeout (start of the next interval - now) the time could still be before the interval. Hence the time-mocking problem.

**Review by stbrody — APPROVED** [review]
lgtm, thanks!

**Comment by stbrody** [comment]
Oh wait, just saw that the test failed on go-ipfs, though it passed in js-ipfs config.  Might be a race condition in the test that fails non-deterministically?  Would be good to look into that and not accidentally introduce a flaky test

**Comment by ukstv** [comment]
Uhums. That's not right. No more flickering tests.

**Pull request merged**